### PR TITLE
SCP-4530 Cannot find block

### PR DESCRIPTION
**Pull request by jhbertra**
- [x] Check in local changes in both NodeClient and Store before making a database query
- [x] Add logs to show when the edge case occurs.

### CODE REVIEW

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs [hunk 1]
```diff
@@ -43,22 +45,24 @@ import Control.Arrow ((&&&))
...
 
 -- | Describes a batch of chain data changes to write.
 data Changes = Changes
-  { changesRollback   :: !(Maybe ChainPoint) -- ^ Point to rollback to before writing any blocks.
-  , changesBlocks     :: ![CardanoBlock]     -- ^ New blocks to write.
-  , changesTip        :: !ChainTip           -- ^ Most recently observed tip of the local node.
-  , changesLocalTip   :: !ChainTip           -- ^ Chain tip the changes will advance the local state to.
-  , changesBlockCount :: !Int                -- ^ Number of blocks in the change set.
-  , changesTxCount    :: !Int                -- ^ Number of transactions in the change set.
+  { changesRollback   :: Maybe ChainPoint -- ^ Point to rollback to before writing any blocks.
```
Comment (bwbush) [inline marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs:60]: Is the dropping of strictness here important?

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs [hunk 2]
```diff
@@ -92,13 +96,14 @@ cost CostModel{..} Changes{..} = changesBlockCount * blockCost + changesTxCount
 
 -- | The set of dependencies needed by the NodeClient component.
 data NodeClientDependencies = NodeClientDependencies
-  { connectToLocalNode    :: !(LocalNodeClientProtocolsInMode CardanoMode -> IO ()) -- ^ Connect to the local node.
-  , getHeaderAtPoint      :: !(GetHeaderAtPoint IO)                                 -- ^ How to load a block header at a given point.
-  , getIntersectionPoints :: !(GetIntersectionPoints IO)                            -- ^ How to load the set of initial intersection points for the chain sync client.
+  -- Explicitly lazy to support value recursion
+  { connectToLocalNode    :: ~(LocalNodeClientProtocolsInMode CardanoMode -> IO ()) -- ^ Connect to the local node.
```
Comment (bwbush) [inline marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs:100]: Is the use of laziness here important for fixing the bug?

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync.hs [hunk 3]
```diff
@@ -49,10 +49,11 @@ data ChainSyncDependencies = ChainSyncDependencies
...
 chainSync :: Component IO ChainSyncDependencies ()
 chainSync = proc ChainSyncDependencies{..} -> do
-  let DatabaseQueries{..} = databaseQueries
-  NodeClient{..} <- nodeClient -< NodeClientDependencies{..}
   let rateLimit = persistRateLimit
-  ChainStore{..} <- chainStore -< ChainStoreDependencies{..}
+  rec
+    let DatabaseQueries{..} = chainStoreDatabaseQueries
+    NodeClient{..} <- nodeClient -< NodeClientDependencies{..}
+    ChainStore{..} <- chainStore -< ChainStoreDependencies{..}
```
Comment (bjornkihlberg) [inline marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync.hs:56]: This is calculating a fix point for some value, right? Presumably the fields in `NodeClient{..}` and `ChainStore{..}`.

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs [hunk 4]
```diff
@@ -166,7 +171,10 @@ pipelinedClient costModel maxCost changesVar getHeaderAtPoint getIntersectionPoi
       -> NumberedChainTip
       -> IO (ClientPipelinedStIdle 'Z NumberedCardanoBlock ChainPoint NumberedChainTip IO ())
     clientStIdle point nodeTip = do
-      clientTip <- fmap blockHeaderToBlockNo <$> runGetHeaderAtPoint getHeaderAtPoint point
+      mClientTip <- (fmap . fmap) blockHeaderToBlockNo <$> runGetHeaderAtPoint getHeaderAtPoint point
+      clientTip <- case mClientTip of
+        Nothing -> error $ "Failed to start chain sync client. Failed to load block number for chain point " <> show point
+        Just tip -> pure tip
```
Comment (bwbush) [inline marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs:177]: On line 185, is `50` an appropriate limit? It seems that would depend upon the PostgreSQL configuration.

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs [hunk 5]
```diff
@@ -244,7 +252,16 @@ mkClientStNext costModel maxCost changesVar getHeaderAtPoint pipelineDecision n
       let clientTip = At blockNo
       pure $ mkClientStIdle costModel maxCost changesVar getHeaderAtPoint pipelineDecision n clientTip tip
   , recvMsgRollBackward = \point tip -> do
-      clientTip <- fmap blockHeaderToBlockNo <$> runGetHeaderAtPoint getHeaderAtPoint point
+      tipFromChanges <- atomically $ lookupPointInChanges point <$> readTVar changesVar
+      clientTip <- case tipFromChanges of
+        Nothing -> do
+          mClientTip <- (fmap . fmap) blockHeaderToBlockNo <$> runGetHeaderAtPoint getHeaderAtPoint point
+          case mClientTip of
+            Nothing -> error $ "Failed to process rollback. Failed to load block number for rollback point " <> show point
```
Comment (bwbush) [inline marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs:260]: So if there is a rollback of the rollback, we might see this error?

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs [hunk 6]
```diff
@@ -244,7 +252,16 @@ mkClientStNext costModel maxCost changesVar getHeaderAtPoint pipelineDecision n
...
+            Just clintTip -> pure clintTip
+        Just clintTip -> do
+          hPutStrLn stderr "Found rollback point in node client changes"
```
Comment (bwbush) [inline marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs:263]: . . . and if there is a rollback that isn't rolled back, we would receive this message?

**Review by bwbush — APPROVED** [review]
I'm testing this on `mainnet` and `preprod` right now. The best bet for testing is too sync, stop syncing for many hours, and then try to sync again.

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs (hunk 2 above)
Comment (jhbertra) [reply marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs:100]: Yes - NodeClient and ChainStore depend recursively on each other now via `ArrowLoop`.

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync.hs (hunk 3 above)
Comment (jhbertra) [reply marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync.hs:56]: Yeah, Exactly - `NodeClientDependencies` depends on a value from `chainStoreDatabaseQueries`, and `ChainStoreDependencies` depends on a value from `NodeClient`

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs (hunk 4 above)
Comment (jhbertra) [reply marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs:177]: It actually depends on the network more - this is the maximum pipelining limit of the chain sync protocol requests, not the database writes. 50 is the value that db-sync uses for this

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs (hunk 5 above)
Comment (jhbertra) [reply marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs:260]: No, most likely not - we'd still be rolling back to a known block.

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs (hunk 6 above)
Comment (jhbertra) [reply marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs:263]: Not necessarily. We'd see this message if we see a rollback while the chain store is persisting a batch of changes, and the node client has a collection of blocks staged for the store to save when it is available.

**Comment by bwbush** [comment]
This version of the code worked correctly when testing on `preprod`:
```console
Nothing to deploy (up-to-date)
Rolling back to block 7e03b0b386fc7696d36522700bc5c05ee79021f1c440e9875d4dc3844f03a5bb at slot 14428952
Saving 16 blocks, 0 transactions. New tip: block 0c2e530d39f72833aea4d3a6bfb47fbf8f9235225b5be5a5c954918c90dcc7cc at slot 14429322 (node tip: 14559440)
Saving 6316 blocks, 1984 transactions. New tip: block 82fab81d9c7303170df50f5abff1f1cf3714c32256dc5dc6caef74aa90be88ce at slot 14559506 (node tip: 14559506)
Found rollback point in node client changes
Found rollback point in node client changes
```

**Review by bwbush — APPROVED** [review]
Yes, much simpler with the map of blocks up to the security parameter.

I need to stop indexing `preprod` for a couple of hours before testing this on chain.

File: marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs [hunk 7]
```diff
@@ -43,9 +42,12 @@ import Control.Arrow ((&&&))
 import Control.Concurrent.Component
 import Control.Concurrent.STM (STM, TVar, atomically, modifyTVar, newTVar, readTVar, writeTVar)
 import Control.Monad (guard)
+import Data.IntMap (IntMap)
+import qualified Data.IntMap as IntMap
 import Data.List (sortOn)
+import Data.Maybe (mapMaybe)
 import Data.Ord (Down(..))
-import Language.Marlowe.Runtime.ChainSync.Database (CardanoBlock, GetHeaderAtPoint(..), GetIntersectionPoints(..))
```
[the remark this answers is not in the document]
Comment (jhbertra) [reply marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/NodeClient.hs:48]: No, none of these database queries are part of the public API.

**Pull request merged**